src/background/EventManager.ts

```diff
@@ -46,34 +28,41 @@ export class EventManager {
       types: ["main_frame"],
     };
     const extraInfoSpec: WebRequest.OnHeadersReceivedOptions[] = ["blocking"];
-    this.webRequestHeadersEmitter.addListener(
-      this.receiveWebRequest.bind(this),
+
+    this.webRequest.onBeforeRequest.addListener(
+      this.resetTabData.bind(this),
+      filter,
+      []
+    );
+    this.webRequest.onHeadersReceived.addListener(
+      this.receiveWebRequestHeaders.bind(this),
       filter,
       extraInfoSpec
     );
-    this.webRequestErrorEmitter.addListener(
+    this.webNavigation.onErrorOccurred.addListener(
       this.receiveWebRequestError.bind(this)
     );
-    this.messageEmitter.addListener(this.receiveMessage.bind(this));
-    this.tabActivatedEmitter.addListener(this.changeBrowserAction.bind(this));
-    this.webRequestBeforeEmitter.addListener(
-      this.resetTabData.bind(this),
-      filter,
-      []
-    );
+    this.runtime.onMessage.addListener(this.receiveMessage.bind(this));
+    this.tabs.onActivated.addListener(this.changeBrowserAction.bind(this));
   }
 
   public resetTabData(requestDetails: { tabId: number }): void {
     this.app.resetTabData(requestDetails.tabId);
   }
 
-  public receiveWebRequest(
+  public async receiveWebRequestHeaders(
     requestDetails: WebRequest.OnHeadersReceivedDetailsType
-  ): void {
-    // using await is not possible here, since making receiveWebRequest async is not allowed
-    this.app.fetchCertificate(requestDetails).then(() => {
-      this.changeBrowserAction(requestDetails);
-    });
+  ): Promise<WebRequest.BlockingResponse> {
+    const hasQualityDecreased = await this.app.fetchCertificate(requestDetails);
+    this.changeBrowserAction(requestDetails);
+
+    if (hasQualityDecreased) {
+      const path = `blocked.html?url=${requestDetails.url}`;
+      this.tabs.create({ url: this.runtime.getURL(path) });
+      return { cancel: true };

```

Comment:
How does the user get redirected? Just asking because i dont get it :sweat:

Reply:
At the moment a new tab with the url `moz-extension://.../blocked.html?url=https://www.google.com/` is created when the quality has decreased that shows the block message. But it is now changed to a redirect.